core/templates/pages/contributor-dashboard-page/services/contribution-and-review.service.ts

```diff
@@ -46,44 +72,102 @@ export class ContributionAndReviewService {
       ContributionAndReviewBackendApiService
   ) {}
 
+  private userCreatedQuestionFetcher: SuggestionFetcher = (
+    new SuggestionFetcher('SUBMITTED_QUESTION_SUGGESTIONS'));
+
+  private reviewableQuestionFetcher: SuggestionFetcher = (
+    new SuggestionFetcher('REVIEWABLE_QUESTION_SUGGESTIONS'));
+
+  private userCreatedTranslationFetcher: SuggestionFetcher = (
+    new SuggestionFetcher('SUBMITTED_TRANSLATION_SUGGESTIONS'));
+
+  private reviewableTranslationFetcher: SuggestionFetcher = (
+    new SuggestionFetcher('REVIEWABLE_TRANSLATION_SUGGESTIONS'));
+
+  /**
+   * Fetches suggestions from the backend.
+   *
+   * @param {SuggestionFetcher} fetcher - The fetcher for a particular
+   *   suggestion type.
+   * @param {boolean} shouldResetOffset - Whether to reset the input fetcher's
+   *   offset to 0 and clear the fetcher's cache. Set this to true to fetch
+   *   results starting from the beginning of all results matching the query.
+   * @returns {Promise<FetchSuggestionsResponse>}
+   */
   private async fetchSuggestionsAsync(
-      fetchType: string
+      fetcher: SuggestionFetcher, shouldResetOffset: boolean
   ): Promise<FetchSuggestionsResponse> {
+    if (shouldResetOffset) {
+      // Handle the case where we need to fetch starting from the beginning.
+      fetcher.offset = 0;
+      fetcher.suggestionIdToDetails = {};
+    }
+    const currentCacheSize: number = Object.keys(
+      fetcher.suggestionIdToDetails).length;
     return (
       this.contributionAndReviewBackendApiService.fetchSuggestionsAsync(
-        fetchType
+        fetcher.type,
+        // Fetch up to two pages at a time to compute if we have more results.
+        // The first page of results is returned to the caller and the second
+        // page is cached.
+        (AppConstants.OPPORTUNITIES_PAGE_SIZE * 2) - currentCacheSize,
```

Comment:
Maybe better to fetch (page_size * 2) - current_cache_size + 1 results, so we know if there's going to be a next page or not?

Reply:
Well, we're already fetching an extra page to determine whether there are "more" results for the caller. The cached second page here is completely invisible to the caller.